[cloudflare] Update caching doc

vicb opened the pull request:
@conico974 @james-elicx Still in draft but let me know if you have feedback, thanks!

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,214 @@ import { Tabs } from "nextra/components";
-`@opennextjs/cloudflare` supports multiple caching mechanisms through a project's OpenNext configuration.
+- An "Incremental Cache" to store the cache data
+- A "Queue" to synchronize and deduplicate revalidations
+- A "Tag Cache" for explicit revalidations via [`revalidateTag`](https://nextjs.org/docs/app/api-reference/functions/revalidateTag) and [`revalidatePath`](https://nextjs.org/docs/app/api-reference/functions/revalidatePath). The Tag Cache is not required for applications using the
```
> using the ....

vicb replied:
Like a Queue using the Tag Cache?

vicb replied:
Oops, good catch!

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,214 @@ import { Tabs } from "nextra/components";
+
+You should use the following implementation for a small site:
+
+- Incremental Cache: use R2 to sore the data
```
```suggestion
- Incremental Cache: use R2 to store the data
```

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,214 @@ import { Tabs } from "nextra/components";
+
+#### Large dynamic site
+
+For a larger dynamic site, you should use the `ShardedDOTagCache` that can handle a higer load than the `D1NextModeTagCache`:
```
```suggestion
For a larger dynamic site, you should use the `ShardedDOTagCache` that can handle a higher load than the `D1NextModeTagCache`:
```

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,214 @@ import { Tabs } from "nextra/components";
+        "class_name": "DOQueueHandler",
+      },
+      {
+        "name": "NEXT_CACHE_DO_SHARDED",
```
```suggestion
        "name": "NEXT_TAG_CACHE_DO_SHARDED",
```

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,214 @@ import { Tabs } from "nextra/components";
+
+#### Large dynamic site
+
+For a larger dynamic site, you should use the `ShardedDOTagCache` that can handle a higer load than the `D1NextModeTagCache`:
```
worth adding a comment about the regional cache on top of r2 here as well?

vicb replied:
Probably.

I want to make something simple first and iterate.

james-elicx commented on pages/cloudflare/caching.mdx:
````diff
@@ -3,17 +3,214 @@ import { Tabs } from "nextra/components";
+
+```ts
+import { defineCloudflareConfig } from "@opennextjs/cloudflare";
+import r2IncrementalCache from "@opennextjs/cloudflare/overrides/incremental-cache/r2-incremental-cache";
````
```suggestion
import r2IncrementalCache from "@opennextjs/cloudflare/overrides/incremental-cache/r2-incremental-cache";
import { withRegionalCache } from "@opennextjs/cloudflare/overrides/incremental-cache/regional-cache";
```

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,214 @@ import { Tabs } from "nextra/components";
+
+#### Staging
+
+For staging, when your site receive low traffic from a single IP, you can replace the memory queue with a DO queue.
```
```suggestion
For staging, when your site receives low traffic from a single IP, you can replace the DO queue with a memory queue.
```

james-elicx commented on pages/cloudflare/caching.mdx:
````diff
@@ -250,19 +446,19 @@ The binding name used in your app's worker is `NEXT_TAG_CACHE_D1`. The `WORKER_S
 ```ts
 // open-next.config.ts
 import { defineCloudflareConfig } from "@opennextjs/cloudflare";
 import kvIncrementalCache from "@opennextjs/cloudflare/kv-cache";
````
```suggestion
import kvIncrementalCache from "@opennextjs/cloudflare/overrides/incremental-cache/kv-incremental-cache";
```

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -250,19 +446,19 @@ The binding name used in your app's worker is `NEXT_TAG_CACHE_D1`. The `WORKER_S
 import kvIncrementalCache from "@opennextjs/cloudflare/kv-cache";
-import d1NextTagCache from "@opennextjs/cloudflare/d1-next-tag-cache";
+import d1NextTagCache from "@opennextjs/cloudflare/overrides/tag-cache/d1-next-tag-cache";
 import memoryQueue from "@opennextjs/cloudflare/memory-queue";
```
```suggestion
import memoryQueue from "@opennextjs/cloudflare/overrides/queue/memory-queue";
```

james-elicx commented on pages/cloudflare/caching.mdx:
```diff
@@ -286,7 +483,7 @@ The service binding should be a self reference to your worker where `<WORKER_NAM
+        "class_name": "DOQueueHandler",
       },
       {
         "name": "NEXT_CACHE_DO_SHARDED",
```
```suggestion
        "name": "NEXT_TAG_CACHE_DO_SHARDED",
```

vicb commented:
Thanks James for catching many errors!

conico974 commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,215 @@ import { Tabs } from "nextra/components";
+
+### Guidelines
+
+#### Small dynamic site
```
I'm not sure if dynamic is the correct word to use here.
By that here we mean a site using ISR and/or On demand revalidation. And SSR works in every case.
We may want to add a small paragraph about it above

conico974 commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,215 @@ import { Tabs } from "nextra/components";
+
+For staging, when your site receives low traffic from a single IP, you can replace the DO queue with a memory queue.
+
+### References
```
Should this be in a different page ?

vicb replied:
Let's keep it this way for now and move the section later

conico974 commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,17 +3,215 @@ import { Tabs } from "nextra/components";
+import doQueue from "@opennextjs/cloudflare/overrides/queue/do-queue";
+
+export default defineCloudflareConfig({
+  incrementalCache: withRegionalCache(r2IncrementalCache, { mode: "short-lived" }),
```
Do we need to add a comment about the `withRegionalCache` or is the reference below enough ?

vicb replied:
Let's keep this section simple.
I think the reference section will need more details - let's add them in follow up PRs

vicb commented:
@conico974 @james-elicx I have rebased and updated the doc, could you please take a look? This will not be the last version but IMO recommendations would be nice to release soonish.
Thanks!

conico974 commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,18 +3,229 @@ import { Tabs } from "nextra/components";
 
-`@opennextjs/cloudflare` supports multiple caching mechanisms through a project's OpenNext configuration.
+- "[Time-based revalidation](https://nextjs.org/docs/app/building-your-application/data-fetching/incremental-static-regeneration#time-based-revalidation)" updates the cache data after the revalidation delay specified by the applications expires
+- "[On-demand revalidation](https://nextjs.org/docs/app/building-your-application/data-fetching/incremental-static-regeneration#on-demand-revalidation-with-revalidatetag)" allows to invalid cache entries with a specific tag (via `revalidateTag`) or at a given path (via `revalidatePath`).
```
```suggestion
- "[On-demand revalidation](https://nextjs.org/docs/app/building-your-application/data-fetching/incremental-static-regeneration#on-demand-revalidation-with-revalidatetag)" allows to invalid cache entries with a specific tag (via `revalidateTag`) or at a given path (via `revalidatePath`). You can also use `res.revalidate` in page router API route.
```

conico974 commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,18 +3,229 @@ import { Tabs } from "nextra/components";
+The `@opennextjs/cloudflare` caching supports rely on 3 components:
+
+- An **Incremental Cache** to store the cache data
+- A **Queue** to synchronize and deduplicate revalidations
```
```suggestion
- A **Queue** to synchronize and deduplicate time-based revalidations
```

conico974 commented on pages/cloudflare/caching.mdx:
```diff
@@ -3,18 +3,229 @@ import { Tabs } from "nextra/components";
+
+#### SSG site
+
+If your site is static, you do not need a Queue nor a Tag Cache. You can also use the `long-lived` mode of the R2 Incremental Cache to make reading from the store faster.
```
I think this one could be confusing for people.
Not sure how we could say it, but we should add something about SSR (probably at the top).
All these recommendations are only for the ISR/SSG and the data cache.
SSR works in every one of these cases.
